include/veriblock/storage/inmem_payloads_provider.hpp

```diff
@@ -93,40 +82,112 @@ struct InmemPayloadsProvider : public PayloadsProvider {
   std::unordered_map<typename T::id_t, std::shared_ptr<T>>& getMap();
 
  private:
-  std::unordered_map<ATV::id_t, std::shared_ptr<ATV>> atvs;
-  std::unordered_map<VTB::id_t, std::shared_ptr<VTB>> vtbs;
-  std::unordered_map<VbkBlock::id_t, std::shared_ptr<VbkBlock>> vbkblocks;
+  std::unordered_map<ATV::id_t, std::shared_ptr<ATV>>& atvs_;
+  std::unordered_map<VTB::id_t, std::shared_ptr<VTB>>& vtbs_;
+  std::unordered_map<VbkBlock::id_t, std::shared_ptr<VbkBlock>>& vbkblocks_;
 };
 
 template <>
 inline std::unordered_map<ATV::id_t, std::shared_ptr<ATV>>&
-InmemPayloadsProvider::getMap() {
-  return atvs;
+InmemPayloadsReader::getMap() {
+  return atvs_;
 }
 template <>
 inline std::unordered_map<VTB::id_t, std::shared_ptr<VTB>>&
-InmemPayloadsProvider::getMap() {
-  return vtbs;
+InmemPayloadsReader::getMap() {
+  return vtbs_;
 }
 template <>
 inline std::unordered_map<VbkBlock::id_t, std::shared_ptr<VbkBlock>>&
-InmemPayloadsProvider::getMap() {
-  return vbkblocks;
+InmemPayloadsReader::getMap() {
+  return vbkblocks_;
 }
 
-template <>
-inline void InmemPayloadsProvider::write(const ATV& v) {
-  auto& m = getMap<ATV>();
-  m.insert({v.getId(), std::make_shared<ATV>(v)});
-  write(v.blockOfProof);
-}
+struct InmemPayloadsWriter : public PayloadsWriter {
```

Comment:
your InmemPayloadsProvider can implement both interfaces, and it would not need to accept maps in constructor

Reply:
fixed